storage/src/index/mod.rs

```diff
@@ -260,117 +255,395 @@ mod tests {
-        iter.remove(); // should be a no-op
-        assert_eq!(*iter.next().unwrap(), 4);
-        iter.remove();
-        iter.remove(); // should be a no-op
-        assert!(context.encode().contains("pruned_total 1"));
+        {
+            let mut cursor = index.get_mut(b"key").unwrap();
+            assert_eq!(*cursor.next().unwrap(), 1);
+            cursor.delete();
+            assert!(!cursor.empty());
+            assert!(context.encode().contains("pruned_total 1"));
+        }
 
         assert_eq!(
-            index.get_iter(b"key").copied().collect::<Vec<_>>(),
-            vec![3, 2, 1]
+            index.get(b"key").copied().collect::<Vec<_>>(),
+            vec![4, 3, 2]
         );
 
-        index.insert(b"key", 4);
+        index.insert(b"key", 1);
         assert_eq!(
-            index.get_iter(b"key").copied().collect::<Vec<_>>(),
-            vec![4, 3, 2, 1]
+            index.get(b"key").copied().collect::<Vec<_>>(),
+            vec![4, 1, 3, 2]
         );
 
         // Test removing from the middle.
-        let mut iter = index.remove_iter(b"key");
-        assert_eq!(*iter.next().unwrap(), 4);
-        assert_eq!(*iter.next().unwrap(), 3);
-        assert_eq!(*iter.next().unwrap(), 2);
-        iter.remove();
-        iter.remove(); // should be a no-op
-        assert!(context.encode().contains("pruned_total 2"));
+        {
+            let mut cursor = index.get_mut(b"key").unwrap();
+            assert_eq!(*cursor.next().unwrap(), 4);
+            assert_eq!(*cursor.next().unwrap(), 1);
+            assert_eq!(*cursor.next().unwrap(), 3);
+            cursor.delete();
+            assert!(!cursor.empty());
+            assert!(context.encode().contains("pruned_total 2"));
+        }
 
         assert_eq!(
-            index.get_iter(b"key").copied().collect::<Vec<_>>(),
-            vec![4, 3, 1]
+            index.get(b"key").copied().collect::<Vec<_>>(),
+            vec![4, 1, 2]
         );
-        index.insert(b"key", 2);
+        index.insert(b"key", 3);
         assert_eq!(
-            index.get_iter(b"key").copied().collect::<Vec<_>>(),
-            vec![2, 4, 3, 1]
+            index.get(b"key").copied().collect::<Vec<_>>(),
+            vec![4, 3, 1, 2]
         );
 
         // Test removing last value.
-        let mut iter = index.remove_iter(b"key");
-        assert_eq!(*iter.next().unwrap(), 2);
-        assert_eq!(*iter.next().unwrap(), 4);
-        assert_eq!(*iter.next().unwrap(), 3);
-        assert_eq!(*iter.next().unwrap(), 1);
-        iter.remove();
-        iter.remove(); // should be a no-op
-        assert!(context.encode().contains("pruned_total 3"));
+        {
+            let mut cursor = index.get_mut(b"key").unwrap();
+            assert_eq!(*cursor.next().unwrap(), 4);
+            assert_eq!(*cursor.next().unwrap(), 3);
+            assert_eq!(*cursor.next().unwrap(), 1);
+            assert_eq!(*cursor.next().unwrap(), 2);
+            cursor.delete();
```

Comment:
can you preserve the test where someone tries to delete/remove twice?

Reply:
Re-added in a separate test, now that it panics to do this.